Add a deal.II-preCICE example (#84)
dealii/code-gallery: coupled_laplace_problem/README.md

Laplace equation coupled to an external simulation program
------------------------------------------
## Overview

preCICE allows to couple deal.II to external simulation software, such as OpenFOAM, SU2, or CalculiX. To keep dependencies of this example minimal, we couple deal.II to an external C++ program, which provides a time varying boundary condition. The deal.II code consists mainly of the [`step-4` tutorial program](https://www.dealii.org/developer/doxygen/deal.II/step_4.html), where a simple Laplace problem is solved. 

Coupling with preCICE is usually carried out along surfaces in order to apply a Dirichlet-Neumann coupling between two domains (volume coupling is also possible). For the sake of simplicity, we couple the external C++ program in a unidirectional fashion to one side of our quadrilateral domain. The external C++ program generates a parabolic boundary profile with time varying amplitude. The boundary values are then used in the Laplace solver as a Dirichlet boundary condition.

## Time discretization

> Coupled simulations deal mostly with time-dependent problems. Hence, we modify the stationary Laplace problem from step-4 by a time-dependent problem

Reviewer: ```suggestion
Coupled simulations deal mostly with time-dependent problems. Hence, we make the stationary Laplace problem from step-4 time dependent.
```

@f{align*}
  \\frac{\partial u}{\partial t}-\Delta u &= f \qquad\qquad & \text{in}\ \Omega,
  \\
  u &= x^2+y^2 \qquad\qquad & \text{on}\ \partial\Omega_s,
  \\
  u &= g(t) \qquad\qquad & \text{on}\ \partial\Omega_c.
@f}
with the fixed Dirichlet boundary \Omega_s, the coupling boundary \Omega_c and the time-dependent coupling data g(t).

The system is consequently discretized by a first-order backward Euler method, resulting in
@f{align*}
  \\frac{u^{n+1}-u^n}{\Delta t}-\Delta u^{n+1} &= f \qquad\qquad & \text{in}\ \Omega,
@f}
at the next time level n+1.

## Requirements

* `deal.II`, version `9.2` or greater. Older versions might work as well, but have not been tested.

> * [preCICE](https://github.com/precice/precice/wiki#1-get-precice), version `2.0` or greater.

Reviewer: There are different options depending on the system. Giving any option here explicitly could lead to the misunderstanding that it is the only way to go. Giving more options would in the end lead to a duplication of the content of the linked page. Just having that link could be the best compromise. What do you think?

## Time discretization
Coupled simulations deal mostly with time-dependent problems. Hence, we make the stationary Laplace problem from step-4 time dependent.
@f{align*}
  \\frac{\partial u}{\partial t}-\Delta u &= f \qquad\qquad & \text{in}\ \Omega,
  \\
  u &= x^2+y^2 \qquad\qquad & \text{on}\ \partial\Omega_s,
  \\
  u &= g(t) \qquad\qquad & \text{on}\ \partial\Omega_c.
@f}
with the fixed Dirichlet boundary \Omega_s, the coupling boundary \Omega_c and the time-dependent coupling data g(t).

The system is consequently discretized by a first-order backward Euler method, resulting in
@f{align*}
  \\frac{u^{n+1}-u^n}{\Delta t}-\Delta u^{n+1} &= f \qquad\qquad & \text{in}\ \Omega,
@f}
at the next time level n+1.

## Requirements

* `deal.II`, version `9.2` or greater. Older versions might work as well, but have not been tested.

* [preCICE](https://github.com/precice/precice/wiki#1-get-precice), version `2.0` or greater. Have a look at the provided link for an installation guide.

## Compiling and running

Similar to the example programs, run
```

> cmake -DDEAL_II_DIR=/path/to/deal.II .

Reviewer 2: To accompany the CMake `HINT` that I suggested earlier, can we also document the `-DPRECICE_DIR` (or whatever you call the CMake variable) here?

```
in this directory to configure the problem.
You can then switch between debug and release mode by calling either
```
make debug
```
or
```
make release
```
This command will generate two executables: one for the `coupled_laplace_problem` and one for the `fancy_boundary_condition` participant.
```
./coupled_laplace_problem
```

> excutes the `coupled_laplace_problem`. In order to start the coupled simulation, execute

Reviewer 2: `excutes` -> `executes`

```
./fancy_boundary_condition
```
in the same directory from another terminal window.

> preCICE and the deal.II solver create during the simulation several log files. In order to remove all result-associated files and clean-up the simulation directory the Allclean script can be executed.

Reviewer 2: `create during the simulation several log files` -> `create several log files during the simulation`
